modules/dnn/include/opencv2/dnn/dnn.hpp

```diff
@@ -406,6 +415,12 @@ CV__DNN_INLINE_NS_BEGIN
                                      std::vector<MatShape> &outputs,
                                      std::vector<MatShape> &internals) const;
 
+        virtual void getTypes(const std::vector<MatType>& inputs,
+                              const int requiredOutputs,
+                              const int requiredInternals,
+                              std::vector<MatType>&outputs,
+                              std::vector<MatType>&internals) const;
```

Comment:
Do we really need such kind of functionality in public API?
What Users scenarios need all types of all layers at once?

It makes sense to support this scenarios instead:
- user set network inputs
- user specify list of requested outputs
- network compiles (internally calls layer's `.finalize()`)
- user asks for layer by its name
- ask types and shapes of layer outputs without passing of any parameters like "inputs/outputs/internals" - pass only index of layer's output.

Note:
- support multiple outputs per layer
- user should not know anything about "internals"
- `getMemoryShapes` / `getLayersShapes()` should be deprecated as primary API too

Reply:
I think it is out of scope of this pull request. I added the new type inference pipeline similar to existing shape inference pipeline because this way produces less architecture changes. I think it will be better to make another pull request with public API modifications.